gocourse05/zooobservation/main.go

```diff
@@ -0,0 +1,129 @@
+/*
+Треба зробити бекенд для сервера, на який надходять нічні зображення з різних камер спостереження з датчиком руху,
+розвішаних по всьому зоопарку. У зоопарку існує кілька типів камер. Деякі камери працюють із зовнішнім світлом, інші —
+в нічному режимі. Треба обробляти дані з різних джерел (типів камер), зберігати в памʼяті історію подій і передавати
+єдиний уніфікований запит на інший сервер. Відповідно, треба зробити кілька типів (структур) які відповідають своїм
+реальним камерам, і декілька інтерфейсів, із якими працює програма. Використовувати контракти й обробляти можливі помилки.
+Також треба написати тести для позитивних і негативних випадків роботи функцій, які оброблюють дані з камер.
+Тут «сервер» — умовна назва для нашої програми.
+Даними може виступати рух певної тварини. Наприклад: тигр, пішов ліворуч; ведмідь, стоїть.
+*/
+package main
+
+import (
+	"fmt"
+	"math/rand/v2"
+	"time"
+)
+
+type Direction string
+
+const (
+	left   Direction = "left"
+	right  Direction = "right"
+	top    Direction = "top"
+	bottom Direction = "bottom"
+)
+
+type HistoryItem struct {
+	time      time.Time
+	direction Direction
+	animalID  int
+}
+
+type History []HistoryItem
+
+type Camera interface {
+	DetectMovement(direction Direction, history *History, animalID int) error
+	SaveToServer(history *History) error
+}
+
+type DayCamera struct {
+	screenshot string
+}
+
+type NightCamera struct {
+	screenshot string
+}
+
+func (d *DayCamera) DetectMovement(direction Direction, history *History, animalID int) error {
```

Comment:
Можемо ось так зробити?
```suggestion
func (d *DayCamera) DetectMovement(direction Direction, historyItems []HistoryItem, animalID int) ([]HistoryItem, error) {
```

Ось як тут https://go.dev/wiki/SliceTricks#move-to-front-or-prepend-if-not-present-in-place-if-possible:
```go
// moveToFront moves needle to the front of haystack, in place if possible.
func moveToFront(needle string, haystack []string) []string {
```

Це такий паттерн, щоб додавати елементи в слайс.